### PR TITLE
AS-2647 firehose and redshift modules

### PR DESCRIPTION
**Pull request by LesBarstow**
Tested using test-module. Valid commands for each: create, facts, delete

### CODE REVIEW

File: cloud/amazon/firehose.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  wait_timeout:
+    description:
+      - The maximum time (in sec) to wait for a delivery stream to become active.
+    required: false
+    default: 300
+requirements:
+    - "python >= 2.7"
+    - "boto3"
+author:
+    - "Les Barstow (@LesBarstow)"
```
Comment (guildencrantz) [inline cloud/amazon/firehose.py]: We shouldn't be putting ownership in the source.

File: cloud/amazon/firehose.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+    for v in required_params:
+        if not module.params.get(v):
+            module.fail_json(msg="Parameter %s required for %s command" % (v, command))
+
+# Mostly for checking 'ACTIVE' state
+def _has_delivery_stream_state(module, conn, state):
+    delivery_stream = _describe_delivery_stream(module, conn)
+
+    if delivery_stream['ResponseMetadata']['HTTPStatusCode'] != 200:
+        return False # Query failed; await_delivery_stream will retry; facts will fail
```
Comment (guildencrantz) [inline cloud/amazon/firehose.py]: Functionality outside this scope should be documented in that scope. Those pieces of code will change, nobody will remember this comment, and then somebody's going to be troubleshooting an issue based on a bad comment.

File: cloud/amazon/firehose.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+    if not delivery_stream:
+        if state == 'DELETED':
+            return True
+        else:
+            return False
+    if not 'DeliveryStreamStatus' in delivery_stream:
+        return False
+    if delivery_stream['DeliveryStreamStatus'] == state:
```
Comment (guildencrantz) [inline cloud/amazon/firehose.py]: This last if/else should just be `return delivery_stream['DelivercyStreamStatus'] == state` so readers don't have to unravel the tautology.

File: cloud/amazon/firehose.py [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+
+def await_delivery_stream_state(module, conn, state):
+    wait_timeout = module.params.get('wait_timeout') + time.time()
+    delivery_stream_name = module.params.get('delivery_stream_name')
+    status = _has_delivery_stream_state(module, conn, state)
+    while wait_timeout > time.time() and not status:
+        time.sleep(5)
+        if wait_timeout <= time.time():
+            module.fail_json(msg="Timeout waiting for Delivery Stream %s" % delivery_stream_name)
+        status = _has_delivery_stream_state(module, conn, state)
```
Comment (guildencrantz) [inline cloud/amazon/firehose.py]: If you want to document the `_has_delivery_stream_state` returns `false` issue, this is where it should be (or in the function description).

File: cloud/amazon/redshift.py [hunk 5]
```diff
@@ -149,15 +147,21 @@
...
+        required: false
+        default: false
+    final_cluster_snapshot_id:
+        description:
+            - A string identifier for a cluster snapshot taken when the cluster is deleted. Required on deletion if skip_final_cluster_snapshot is false
+        required: false
+        default: null
 extends_documentation_fragment:
     - aws
     - ec2
```
Comment (guildencrantz) [inline cloud/amazon/redshift.py]: Author attributions should not be in source.

**Comment by guildencrantz** [comment]
:-1:

**Comment by guildencrantz** [comment]
:+1:

**Comment by trinitronx** [comment]
@LesBarstow, @guildencrantz :  Just fixed merge conflicts & merged!

**Pull request merged**